File: kotlin/rayfold-core/src/main/kotlin/dev/rayfold/core/Policy.kt
package dev.rayfold.core

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Policy evaluation (spec/06). Absent allow = allowed; deny after allow; viewer-dependent denial without a viewer =
 * unauthenticated. An expression that cannot be evaluated ([ExprError]) fails closed: a broken allow does not admit
 * and a broken deny denies.
 */
object Policy {
    enum class Decision { ALLOW, DENY, UNAUTHENTICATED }

    private fun expr(annotations: List<Annotation>, name: String, mode: String): JsonObject? =
        annotations.find(name)?.args?.get(mode).exprOrNull()

    fun hasPolicy(annotations: List<Annotation>, mode: String): Boolean =
        expr(annotations, "allow", mode) != null || expr(annotations, "deny", mode) != null

    fun decide(annotations: List<Annotation>, mode: String, env: ExprEnv): Decision {
        val allow = expr(annotations, "allow", mode)
        val deny = expr(annotations, "deny", mode)
        if (allow == null && deny == null) return Decision.ALLOW
        fun denied(e: JsonObject) = if (env.viewer is JsonNull && Expr.referencesViewer(e)) Decision.UNAUTHENTICATED else Decision.DENY
        fun holds(e: JsonObject, onError: Boolean) = try { Expr.eval(e, env) == JsonPrimitive(true) } catch (x: ExprError) { onError }
        if (allow != null && !holds(allow, onError = false)) return denied(allow)
        if (deny != null && holds(deny, onError = true)) return denied(deny)
        return Decision.ALLOW
    }

    /**
     * The part of a type's read policy a data source can apply itself (spec 06 section 4), or null when nothing can
     * be pushed. A `deny` needs the runtime's own check, so nothing is pushed when one is present.
     */
    fun pushableFilter(annotations: List<Annotation>): JsonObject? {
        if (expr(annotations, "deny", "read") != null) return null
        val allow = expr(annotations, "allow", "read") ?: return null
        return if (isPushable(allow)) allow else null
    }

    /** Mirrors isPushable in packages/schema/src/expr.ts: what a data source can evaluate without the row's shape. */
    fun isPushable(e: JsonObject): Boolean = when (e["k"]?.jsonPrimitive?.content) {
        "lit" -> true
        // a path into the row itself is pushable only when it names one of its own columns
        "path" -> (e["root"]?.jsonPrimitive?.content ?: "this") != "this" || (e["path"] as? JsonArray)?.size == 1
        "not" -> (e["e"] as? JsonObject)?.let { isPushable(it) } == true
        "bin" -> (e["l"] as? JsonObject)?.let { isPushable(it) } == true && (e["r"] as? JsonObject)?.let { isPushable(it) } == true
        "list" -> (e["items"] as? JsonArray)?.all { it is JsonObject && isPushable(it) } == true
        "call" -> e["fn"]?.jsonPrimitive?.content != "now" && (e["args"] as? JsonArray)?.all { a ->
            a is JsonObject && isPushable(a) && !((a["k"]?.jsonPrimitive?.content == "path") && (a["root"]?.jsonPrimitive?.content ?: "this") == "this")
        } == true
        else -> false
    }

    fun error(d: Decision, what: String): RayfoldException = if (d == Decision.UNAUTHENTICATED)
        RayfoldException(Code.UNAUTHENTICATED, "Sign in to access $what")
    else RayfoldException(Code.PERMISSION_DENIED, "Not allowed to access $what")
}

File: packages/server/src/policy.ts
/** Policy evaluation. Spec: spec/06-auth.md. */
import { evalExpr, isPushable, referencesViewer, type Annotation, type Expr, type ExprEnv } from "@rayfold/schema";
import { RayfoldError } from "./protocol.ts";

export type PolicyMode = "read" | "write";

interface Policies {
  allow?: Expr;
  deny?: Expr;
}

const cache = new WeakMap<Annotation[], Record<PolicyMode, Policies>>();

export function policiesOf(annotations: Annotation[], mode: PolicyMode): Policies {
  let entry = cache.get(annotations);
  if (!entry) {
    entry = { read: extract(annotations, "read"), write: extract(annotations, "write") };
    cache.set(annotations, entry);
  }
  return entry[mode];
}

function extract(annotations: Annotation[], mode: PolicyMode): Policies {
  const out: Policies = {};
  for (const a of annotations) {
    if (a.name !== "allow" && a.name !== "deny") continue;
    const v = a.args[mode];
    if (v && typeof v === "object" && "$expr" in v) out[a.name] = (v as { $expr: Expr }).$expr;
  }
  return out;
}

export function hasPolicy(annotations: Annotation[], mode: PolicyMode): boolean {
  const p = policiesOf(annotations, mode);
  return p.allow !== undefined || p.deny !== undefined;
}

export type Decision = "allow" | "deny" | "unauthenticated";

/** Absent allow = allowed; deny evaluated after allow. Viewer-dependent denial with no viewer = unauthenticated. */
export function decide(annotations: Annotation[], mode: PolicyMode, env: ExprEnv): Decision {
  const p = policiesOf(annotations, mode);
  if (p.allow === undefined && p.deny === undefined) return "allow";
  const denied = (e: Expr) => (env.viewer === null || env.viewer === undefined) && referencesViewer(e) ? "unauthenticated" : "deny";
  // An expression that cannot be evaluated (say, ordering a string against a boolean) fails closed: it never allows
  // and always denies.
  const holds = (e: Expr): boolean | "error" => {
    try {
      return evalExpr(e, env) === true;
    } catch {
      return "error";
    }
  };
  if (p.allow !== undefined && holds(p.allow) !== true) return denied(p.allow);
  if (p.deny !== undefined && holds(p.deny) !== false) return denied(p.deny);
  return "allow";
}

export function decisionError(d: Exclude<Decision, "allow">, what: string): RayfoldError {
  return d === "unauthenticated"
    ? new RayfoldError("unauthenticated", `Sign in to access ${what}`)
    : new RayfoldError("permission_denied", `Not allowed to access ${what}`);
}

/** The read policy of a type if it can be pushed down to a loader as a filter (spec 06 §4). */
export function pushableFilter(annotations: Annotation[]): Expr | undefined {
  const p = policiesOf(annotations, "read");
  if (p.deny !== undefined) return undefined; // deny needs post-filter semantics
  if (p.allow !== undefined && isPushable(p.allow)) return p.allow;
  return undefined;
}
